### PR TITLE
[fix] rewrite Yahoo-News engine

### PR DESCRIPTION
**Pull request by return42**
Many things have been changed since last review of this engine.  This patch fix xpath selectors and is a complete review / rewrite of the engine.

## Why is this change important?

Yahoo News engine does not work anymore

## How to test this PR locally?

     !yhn linux

### CODE REVIEW

File: searx/utils.py [hunk 1]
```diff
@@ -607,3 +607,30 @@ def eval_xpath_getindex(elements, xpath_spec, index, default=NOTSET):
...
     return default
+
+
+class RE:
+    u"""regular expression that stores last match (like Perl's ``=~`` operator)"""
+
+    # pylint: disable=missing-function-docstring
+    def __init__(self, *args, **kwargs):
+        self.re = re.compile(*args, **kwargs)
+        self.last_match = None
```
Comment (dalf) [inline searx/utils.py]: In the current architecture (multi thread), it creates side effect:
* the RE instance is shared in the yahoo engine between the different requests
* two requests can create:
     * write from thread 1
     * write from thread 2
     * read from thread 2
     * read from thread 1  <--- error

With async everywhere and no thread, it is technically okay, but still.

File: searx/utils.py [hunk 2]
```diff
@@ -607,3 +607,30 @@ def eval_xpath_getindex(elements, xpath_spec, index, default=NOTSET):
...
+
+    def search(self, *args, **kwargs):
+        self.last_match = self.re.search(*args, **kwargs)
+        return self.last_match
+
+    def __getattr__(self, attr):
+        return getattr(self.re, attr)
+
+    def __getitem__(self, group):
+        if group < 0 or group > self.groups - 1:
```
Comment (dalf) [inline searx/utils.py]: `self.groups` is never assigned

File: searx/utils.py (hunk 1 above)
Comment (return42) [reply searx/utils.py]: Its a class instance like any other .. should I move the `RE_ago` into the `response` function?

https://github.com/searx/searx/pull/2640/files#diff-94618cb260d81eafbc56387ca5bc868415b94871879888e4c90f2c5f2049bc63R55

File: searx/engines/yahoo_news.py [hunk 3]
```diff
@@ -22,90 +40,77 @@
...
+
+        item = {
+            'url': url,
+            'title': title,
+            'content': content,
+            'img_src' : img_src
+        }
+
+        pub_date = extract_text(result.xpath('.//span[contains(@class,"s-time")]'))
+        if RE_ago.search(pub_date):
```
Comment (dalf) [inline searx/engines/yahoo_news.py]: I don't know perl, but I'm not sure to understand the `RE` class.

The code here seems to be identical to:
```python
ago = regex_ago.search(pub_date)
if ago:
	ago = int(ago.group(0))
	if ago.group(1) == 'minute':
		pub_date = datetime.now() - timedelta(minutes=c)
       ....
```

Even simpler
```python
AGO_TIMEDELTA = {
  'minute': timedelta(minutes=1),
  'hour': timedelta(hour=1),
  'day': timedelta(day=1),
  'month': timedelta(month=1),
  'year': timedelta(year=1),
}

...
raw_pub_date= regex_ago.search(pub_date)
if ago:
	number = int(raw_pub_date.group(0))
	delta = AGO_TIMEDELTA[raw_pub_date.group(1)]
        pub_date = datetime.now() - delta * number
else:
        try:
            pub_date = parser.parse(raw_pub_date)
        except parser.ParserError:
            pub_date = None
```

**Review by dalf — APPROVED** [review]
Thank you!

**Pull request merged**